### PR TITLE
Time for upgrade

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
The structure of the app is now compatible with Android Studio/Gradle 2.3.0 
Also added two listeners for changing wifi/data state, which can work independently of the app itself.
One can now execute

```bash
adb shell am broadcast -a io.appium.settings.wifi --es setstatus enable
```

or

```bash
adb shell am broadcast -a io.appium.settings.wifi --es setstatus disable
```

to enable/disable Wi-Fi state on the device under test.

AND

```bash
adb shell am broadcast -a io.appium.settings.data_connection --es setstatus enable
```

or

```bash
adb shell am broadcast -a io.appium.settings.data_connection --es setstatus disable
```

to enable/disable data connection state on the device under test.

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
Location mocking has still not been tested, since I need more detailed instructions on how to do it properly. But this change will for sure require some additions in Appium code after it is merged.

**Comment by imurchie** [comment]
The first issue is that this still tries the `ant` build command on `npm install`. 

```
isaac-2 io.appium.settings → git mykola-mokhnach-add_listeners → npm install                                                                            4.7.2
npm WARN prepublish-on-install As of npm@5, `prepublish` scripts will run only for `npm publish`.
npm WARN prepublish-on-install (In npm@4 and previous versions, it also runs for `npm install`.)
npm WARN prepublish-on-install See the deprecation note in `npm help scripts` for more information.

> io.appium.settings@2.1.0 prepublish /Users/isaac/code/io.appium.settings
> ant clean debug

Buildfile: /Users/isaac/code/io.appium.settings/build.xml

BUILD FAILED
/Users/isaac/code/io.appium.settings/build.xml:49: Source resource does not exist: /Users/isaac/code/io.appium.settings/project.properties

Total time: 0 seconds

npm WARN io.appium.settings@2.1.0 license should be a valid SPDX license expression
npm ERR! Darwin 16.4.0
npm ERR! argv "/Users/isaac/.nvm/versions/node/v4.7.2/bin/node" "/Users/isaac/.nvm/versions/node/v4.7.2/bin/npm" "install"
npm ERR! node v4.7.2
npm ERR! npm  v4.1.1
npm ERR! code ELIFECYCLE
npm ERR! io.appium.settings@2.1.0 prepublish: `ant clean debug`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the io.appium.settings@2.1.0 prepublish script 'ant clean debug'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the io.appium.settings package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ant clean debug
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs io.appium.settings
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls io.appium.settings
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/isaac/code/io.appium.settings/npm-debug.log
```

**Comment by mykola-mokhnach** [comment]
@imurchie Fixed with the recent commit

**Comment by imurchie** [comment]
Cool. Giving this a whirl locally.

File: index.js [hunk 1]
```diff
@@ -3,5 +3,5 @@
 var path = require('path');
 
 module.exports = {
-  path: path.resolve(__dirname, "bin", "settings_apk-debug.apk")
+  path: path.resolve(__dirname, 'app', 'build', 'outputs', 'apk', 'settings_apk-debug.apk');
```
Comment (imurchie) [inline index.js]: Semi-colon at the end is in error.

**Comment by mykola-mokhnach** [comment]
@imurchie We will need to build this app using the same key, which was used to build the previous version, otherwise application update with 'install -r' will fail.

**Comment by imurchie** [comment]
What do you mean by "key"?

**Comment by mykola-mokhnach** [comment]
I mean the signing key

**Comment by mykola-mokhnach** [comment]
It looks like changing the sate of mobile data is not possible without root permissions on Android 5.0+
Added a note about this into README

**Comment by vrunoa** [comment]
@mykola-mokhnach @imurchie can we add special cases for enabling/disabling wifi(data) for emulators ? I'm trying this app and it's not changing the network state. I can make a PR in `appium-adb`. 
For emus we can use `adb shell svc data disable`

File: app/build.gradle [hunk 2]
```diff
@@ -0,0 +1,36 @@
+apply plugin: 'com.android.application'
+
+android {
+    compileSdkVersion 23
+    buildToolsVersion '25.0.0'
+
+    defaultConfig {
+        minSdkVersion 17
+        targetSdkVersion 23
```
Comment (rkavalap) [inline app/build.gradle]: If I am not mistaken targetSdkVersion needs to be 25 if we want this for nougat as well

Comment (mykola-mokhnach) [inline app/build.gradle]: not necessary. 23 is enough to advertise we are compatible with new permissions setting logic.

**Comment by mykola-mokhnach** [comment]
@vrunoa See http://stackoverflow.com/questions/4008627/how-to-put-an-avd-emulator-into-offline-mode
svc command execution works from external shell only, but not from the device. That is why it should be better to implement this directly in appium-adb

**Comment by imurchie** [comment]
On that idea (needing to do some stuff through adb, some on the device) at some point it would be nice to transform this from just a basic app that we install, and import into the project so we can get the path, to a full-fledged node package that includes an API for interacting with settings, and could do both sorts of operations.

**Comment by vrunoa** [comment]
@mykola-mokhnach yeah, that was my proposal in `appium-adb` use this helper app only for real devices and use svc command for emulators.

**Comment by rkavalap** [comment]
👍

**Comment by egorvas** [comment]
Waiting for merge this PR as soon as possible

**Comment by imurchie** [comment]
I'm not sure how to test whether this works, so I'll defer to others. The code itself looks good to me.

**Comment by mykola-mokhnach** [comment]
@imurchie Thanks for the review
Can you please do a quick check for the recent commit? I've fixed the license comment and cleaned up obsolete stuff.

**Comment by imurchie** [comment]
Looks good to me.

**Comment by jlipps** [comment]
published as `io.appium.settings@2.2.0`

**Pull request merged**